### PR TITLE
Enhance missile collision in SpawnLightning

### LINKED ISSUE #4384: Chain lightning can shoot through walls
Happens when there are enemies behind the wall:

image

Save: [single_1.zip](https://github.com/diasurgical/devilutionX/files/8459317/single_1.zip)

### PR DESCRIPTION
**Pull request by obligaron**
Fixes #4384

Notes:
- After investigating this is the same problem as #2837.
- Extract fix from #2831 in new function (first commit). 
- Used the same logic for `SpawnLightning`
- `SpawnLightning` wasn't converted to `MoveMissileAndCheckMissileCol`, cause it doesn't call `CheckMissileCol`. 
- The same problem could occur in other places `UpdateMissilePos` is used (for example rhino). This is not addressed with this pr.

### CODE REVIEW

**Comment by obligaron** [comment]
@AJenbo could you take a look?
If this is merged, I could use the new missile helpers introduced in #4632 (separate pr, cause i also want to change `scrollrt.cpp`). 🙂 
Else I would get a merge conflict. 😉

File: Source/missiles.cpp [hunk 1]
```diff
@@ -506,38 +500,65 @@ void MoveMissileAndCheckMissileCol(Missile &missile, int mindam, int maxdam, boo
...
+		if (ignoreStart && missile.position.start == tile)
+			return true;
+
+		CheckMissileCol(missile, mindam, maxdam, false, tile, false);
+
+		// Did missile hit anything?
+		if (missile._mirange != 0)
+			return true;
+
+		if (missile._miHitFlag && MissilesData[missile._mitype].MovementDistribution == MissileMovementDistrubution::Blockable)
```
Comment (StephenCWills) [inline Source/missiles.cpp:545]: It's not relevant to this PR, but I just noticed the typo in MissileMovementDistr<ins>**u**</ins>bution.

Comment (obligaron) [reply Source/missiles.cpp]: Sorry for that one. 😳 
I fixed it in a separate commit.
if you think this should be better in another pr, I can drop the commit.

File: Source/missiles.cpp [hunk 2]
```diff
@@ -477,16 +477,10 @@ void MoveMissileAndCheckMissileCol(Missile &missile, int mindam, int maxdam, boo
...
-	if (possibleVisitTiles == 0) {
-		// missile didn't change the tile... check that we perform CheckMissileCol only once for any monster/player to avoid multiple hits for slow missiles
-		if (missile.lastCollisionTargetHash == tileTargetHash)
-			return;
-	}
-	// remember what target CheckMissileCol was checked against
-	missile.lastCollisionTargetHash = tileTargetHash;
-	// Did the missile skipped a tile?
+	// Did the missile skip a tile?
 	if (possibleVisitTiles > 1) {
```
[the remark this answers is not in the document]
Comment (obligaron) [reply Source/missiles.cpp:472]: This is a optimization for the special case `possibleVisitTiles == 1`.
When a missile is known to have moved only one tile, we don't need to search for skipped/in-between tiles.
```
			// we are at the original calculated position => resume with normal logic
			if (tile == missile.position.tile)
				break;
```
Will end the loop. But we could have done multiple iterations to get to this point (moving away from original tile can take more then one iteration).

If we want to remove this optimization
`		auto incVelocity = missile.position.velocity * (0.01f / (float)(possibleVisitTiles - 1));`
need to be adjusted to not divide by zero if `possibleVisitTiles == 1`. But this is simple.

[the remark this answers is not in the document]
Comment (StephenCWills) [reply Source/missiles.cpp:472]: Oh, you're right. I mistakenly interpreted this as `possibleVisitTiles > 0`.

**Review by StephenCWills — APPROVED** [review]

**Pull request merged**